### ISSUE TITLE
[ban-types] recommended replacement for any object does not work for interfaces

### ISSUE THREAD

**Issue by steinybot** [typescript-eslint/typescript-eslint#3291]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/<rule>": ["<setting>"]
  }
}
```

(I dunno, whatever the defaults are for the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)).

```TS
interface A {
  a: string;
}

const a: A = {
  a: "A.a",
};

const empty: {} = a;

// Type 'A' is not assignable to type 'Record<string, never>'.
//   Index signature is missing in type 'A'.ts(2322)
const bad: Record<string, never> = a;

const good: Record<never, never> = a;
```

**Expected Result**

```
// If you want a type meaning "empty object", you probably want `Record<never, never>` instead.
const empty: {} = a;
```

**Actual Result**

```
// If you want a type meaning "empty object", you probably want `Record<string, never>` instead.
const empty: {} = a;
```

**Additional Info**

[Index signature is missing in type (only on interfaces, not on type alias)
](https://github.com/microsoft/TypeScript/issues/15300)
[\[ban-types\] Should not ban {} by default!](https://github.com/typescript-eslint/typescript-eslint/issues/2063)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | ??? |
| `@typescript-eslint/parser`        | ??? |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | ??? (extension version 2.1.20) |
| `node`                             | `14.15.5` |

**Comment by bradzacher** [comment]
In your example you're assigning an object with properties to the suggested "empty object type" and it errors.
So this is working correctly...

You're looking for a type that means "empty object". Which means you want a type which BLOCKS you from assigning a non-empty object to it. `Record<string, never>` is that type.

Unless I'm missing something?

**Comment by steinybot** [comment]
I don't think having additional properties should prevent assignment.

If that were the case then this wouldn't work:
```
interface A {
  a: string;
}

interface B {
  a: string;
  b: string;
}

const b: B = {
  a: "B.a",
  b: "B.b",
};

const a: A = b;
```

Applying the Liskov substitution principle, any object should be assignable to the empty object.

**Comment by steinybot** [comment]
I think what I am referring to as the empty object is what the error message calls the "any object". (Not sure why you would want to prevent assignment to an empty object but that's another matter).

So the revised issue is:

```
interface A {
  a: string;
}

const a: A = {
  a: "A.a",
};

const empty: {} = a;

// Type 'A' is not assignable to type 'Record<string, unknown>'.
//   Index signature is missing in type 'A'.ts(2322)
const bad: Record<string, unknown> = a;

const good: Record<never, never> = a;
```

expected:

```
// If you want a type meaning "any object", you probably want `Record<never, never>` instead.
const empty: {} = a;
```

actual:

```
// If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
const empty: {} = a;
```

**Comment by bradzacher** [comment]
This is just a limitations of TypeScript's `interface`s unfortunately - and there's no good workaround.
https://github.com/microsoft/TypeScript/issues/15300

The issue with using `never` as the key is that then your object is unindexable:
```ts
declare const x: Record<never, never>;

x['a']
// Element implicitly has an 'any' type because expression of type '"a"' can't be used to index type 'Record<never, never>'.
//  Property 'a' does not exist on type 'Record<never, never>'.(7053)
```

Which means that it becomes unusable without asserting its type back into something usable.

**Comment by steinybot** [comment]
It turns out that indexing it is easy with the correct type ascription:
```
declare const x: Record<never, never>;
const x2: Record<string, unknown> = x;

x2["a"];
```

**Comment by bradzacher** [comment]
yeah - but that's the point.

For example - if you declare a function which accepts any object your function would be:
```ts
function foo(arg: Record<never, never>) {
  const argRecast: Record<string, unknown> = arg;
  // ...
}
```

Which is cumbersome and just bad code - hence the suggestion is to use `Record<string, unknown>`.
The hope is that they make interfaces easier to work with in regards to index signatures.

The problem is that I don't think there is a way to assert an interface to an index signature without asserting it through unknown/any.

Using a different type like `object` works, but `object` is looser as it allows arrays to be assigned (as they are technically objects). Looking at it as well - `Record<never, never>` actually does the same thing! Which is undesirable for most cases.

`Record<string, unknown>` is the only type that sort of works, it's just interfaces which ruin it.

If there is a better type I'd happily switch to it - but for now, it's probably the best option.
